src/runtime/backend_loader.cpp

```diff
@@ -0,0 +1,178 @@
+ *    and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+
+#include "hipSYCL/runtime/backend_loader.hpp"
+#include "hipSYCL/common/debug.hpp"
+#include "hipSYCL/common/config.hpp"
+
+#include <cassert>
+#include <filesystem>
+#include <dlfcn.h>
+
+namespace {
+
+void close_plugin(void *handle) {
+  int err = dlclose(handle);
+
+  if (err != 0) {
+    HIPSYCL_DEBUG_ERROR << "backend_loader: dlclose() failed" << std::endl;
+  }
+}
+
+bool load_plugin(const std::string &filename, void *&handle_out,
+                 std::string &backend_name_out) {
+  void *handle = dlopen(filename.c_str(), RTLD_NOW);
+
+  if (!handle) {
+    HIPSYCL_DEBUG_ERROR << "backend_loader: Could not load backend plugin: "
+                        << filename << std::endl;
+    char *err = dlerror();
+    if (err) {
+      HIPSYCL_DEBUG_ERROR << err << std::endl;
+    }
+
+    return false;
+  } else {
+    void *symbol = dlsym(handle, "hipsycl_backend_plugin_get_name");
+
+    char *err = dlerror();
+    if (err) {
+      HIPSYCL_DEBUG_ERROR << "backend_loader: Could not retrieve backend name"
+                          << err << std::endl;
+    }
+    else {
+      auto get_name =
+          reinterpret_cast<decltype(&hipsycl_backend_plugin_get_name)>(symbol);
+
+      handle_out = handle;
+      backend_name_out = get_name();
+
+      return true;
+    }
+  }
+
+  // In case of failure, close handle to avoid resource leak
+  close_plugin(handle);
+  return false;
+}
+
+hipsycl::rt::backend *create_backend(void *plugin_handle) {
+  assert(plugin_handle);
+
+  void *symbol = dlsym(plugin_handle, "hipsycl_backend_plugin_create");
+  char *err = dlerror();
+  if (err) {
+    HIPSYCL_DEBUG_ERROR
+        << "backend_loader: Could not find symbol for backend creation" << err
+        << std::endl;
+    
+    return nullptr;
+  }
+  
+  auto create_backend_func =
+      reinterpret_cast<decltype(&hipsycl_backend_plugin_create)>(symbol);
+
+  return create_backend_func();
+}
+
+}
+
+namespace hipsycl {
+namespace rt {
+
+void backend_loader::query_backends() {
+  std::string install_prefix = HIPSYCL_INSTALL_PREFIX;
```

Comment:
First priority is making sure that current packaging strategies with fixed install paths work as expected. Redistributing is a different story and might also require changes to the way we use `-rpath`. I see the use case. Not sure if it's a good idea to just randomly start loading shared libraries from paths that are not necessarily trusted if something doesn't work.

The better approach is probably to introduce a field for the `settings` infrastructure that can be set via environment variable to override this. But I'm not sure if this might not be better a PR in itself.